kubevela/kubevela
# Feat: New Component definition for Statefulset

### Description of your changes

This PR adds a new component definition, `statefulservice` to deploy a long-running, scalable application with guarantees about the ordering and uniqueness of the replicas


Fixes #

I have:


### How has this code been tested

The PR includes a test that deploys an application with a `statefulservice` component in `./test/e2e-test/application_test.go` file
```
...
It("Test stateful app with two replicas", func() {
   applyApp("app_stateful.yaml")
   By("Apply the application rollout go directly to the target")
   verifyStatefulsetRunningExpected("hello-world", 2, "crccheck/hello-world")
   verifyComponentRevision("hello-world", 1)
})
...
```
so it would be enough to run the tests:
```
$ make e2e-test
```

To test manually:

1. Add the component definition in the `vela-system` namespace
    ```
    kubectl -n vela-system create -f ./charts/vela-core/templates/defwithtemplate/statefulservice.yaml
    componentdefinition.core.oam.dev/statefulservice created
    ```
2. Deploy the application included in the tests:
    ````
    kubectl create -f test/e2e-test/testdata/app/app_stateful.yaml 
   application.core.oam.dev/statefulservice-replicas created
    ````
3. Check the statefulset generated
    ````
    kubectl get statefulsets.apps
    NAME          READY   AGE
    hello-world   2/2     67s

    k get pods
    NAME                          READY   STATUS             RESTARTS   AGE 
    hello-world-0                 1/1     Running            0          94s
    hello-world-1                 1/1     Running            0          92s
    ````





### Special notes for your reviewer
